oauthenticator/okpy.py

```diff
@@ -0,0 +1,109 @@
+    _OAUTH_AUTHORIZE_URL =  "https://okpy.org/oauth/authorize"
+    def get(self):
+        self.authorize_redirect(
+            redirect_uri = self.authenticator.oauth_callback_url,
+            client_id = self.authenticator.client_id,
+            scope= ['all'],
+            response_type = 'code')
+
+class OkpyCallbackHandler(BaseHandler, OAuth2Mixin):
+    """ Basic handler for Okpy callback.
+        Calls authenticator to verify user and saves the state in user.
+    """
+    def get_auth_request(self, code):
+        params = dict(
+            redirect_uri = self.authenticator.oauth_callback_url,
+            code = code,
+            grant_type = 'authorization_code'
+        )
+        b64key = base64.b64encode(
+            bytes("{}:{}".format(self.authenticator.client_id,
+                                 self.authenticator.client_secret), "utf8"
+                 )
+        )
+        url = url_concat(OAUTH_ACCESS_TOKEN_URL, params)
+        req = HTTPRequest(url,
+                method = "POST",
+                headers = { "Accept": "application/json",
+                            "User-Agent": "JupyterHub",
+                            "Authorization": "Basic {}".format(b64key.decode("utf8"))},
+                            body = '' # Body is required for a POST...
+                          )
+        return req
+
+    @gen.coroutine
+    def get(self):
+        username, state = yield self.authenticator.authenticate(self, None)
+        if username:
+            user = self.user_from_username(username)
+            user.auth_state = state
+            self.db.add(user)
+            self.db.commit()
+            self.set_login_cookie(user)
+            self.redirect(url_path_join(self.hub.server.base_url, 'home'))
+        else:
+            raise web.HTTPError(403, "Authentication Failed.")
+
+    def get_user_info_request(self, access_token):
+        headers = {"Accept": "application/json",
+                   "User-Agent": "JupyterHub",
+                   "Authorization": "Bearer {}".format(access_token)}
+        params = {"envelope" : "false"}
+        url = url_concat(OKPY_USER_URL, params)
+        req = HTTPRequest(url, method = "GET", headers = headers)
+        return req
+
+class OkpyOAuthenticator(OAuthenticator, OAuth2Mixin):
+    login_service = "Okpy"
+    callback_handler = OkpyCallbackHandler
+    login_handler = OkpyLoginHandler
+    @gen.coroutine
+    def authenticate(self, handler, data = None):
+        code = handler.get_argument("code", False)
+        if not code:
+            raise web.HTTPError(400, "Authentication Cancelled.")
+        http_client = AsyncHTTPClient()
+        auth_request = handler.get_auth_request(code)
+        response = yield http_client.fetch(auth_request)
+        if not response:
+            self.clear_all_cookies()
+            raise HTTPError(500, 'Authentication Failed: Token Not Acquired')
+        state = json.loads(response.body.decode('utf8', 'replace'))
```

Comment:
Where else can we store them? This is the only place we can store them that survives hub restarts and user container restarts. If the db is blown away, that's ok - it gets repopulated on login. I don't think there's any way other than this, and there's other people wanting to do the exact same thing in https://github.com/jupyterhub/jupyterhub/issues/1063